File: minion/files/src/stats.py
#!/usr/bin/env python3

#
# Author: Davi Gupta (davigupta@gmail.com), Sep 2020
#

"""
    Send dropped packet do null pod for further inspection and generating stats
"""
import asyncio
import re

OPENED = False
MYW = 0

async def pak_drop(pak, reason, log):
    global OPENED, MYW

    log.error("packet drop: {}".format(reason))
    top, bottom = pak.split(b'\r\n\r\n', 1)
    pak = top +  b'\r\n' + b'x-nextensio-drop: ' + reason.encode('utf-8') + b'\r\n\r\n'
    pak = re.sub(b"content-length: .*", b"content-length: 0", pak, re.MULTILINE)
    if not OPENED:
        try:
            _, MYW = await asyncio.open_connection("blackhole", 10000)
            OPENED = True
        except :
            return
    MYW.write(pak)
    try:
        await MYW.drain()
    except (ConnectionResetError, ConnectionAbortedError) as e:
        MYW.close()
        OPENED = False

async def run_test(count, log):
    pak = b"GET / HTTP/1.1\r\nHost: gateway.sjc.nextensio.net\r\nuser-agent: shorty\r\nx-nextensio-for: 127.0.0.1\r\nx-nextensio-uuid: 123\r\ncontent-length: 23\r\n\r\n<body>\r\nhello\r\n<\body>"
    for i in range(0, count):
        await pak_drop(pak, "access-denied", log)

if  __name__ == "__main__":
    import sys
    import logging
    logger = logging.getLogger('stats')
    logging.basicConfig(stream=sys.stdout, level=logging.DEBUG)
    try:
        task = run_test(2, logger)
        asyncio.ensure_future(task)
        asyncio.get_event_loop().run_forever()
    finally:
        logger.info("exiting")
        print("exiting")

File: nxts.io/minion.io/stats/stats.go
/*
 * stats.go: Packet statistics, e.g., drop
 *
 * Author: Davi Gupta (davigupta@gmail.com), Sep 2020
 */

package stats

import (
	"bytes"
	"go.uber.org/zap"
	"net"
	"regexp"
	"strconv"
	"strings"
)

var nullConn net.Conn = nil

func PakDrop(pak []byte, reason string, s *zap.SugaredLogger) {
	s.Debugf("stats: packet drop due to %v\n", reason)
	if nullConn == nil {
		portStr := strconv.Itoa(10000)
		servAddr := strings.Join([]string{"127.0.0.1", portStr}, ":")
		conn, e := net.Dial("tcp", servAddr)
		if e != nil {
			s.Debugw("stats:", "error", e)
			return
		}
		nullConn = conn
	} else {
		s.Debug("stats: connection exists\n")
	}
	z := bytes.SplitN(pak, []byte("\r\n\r\n"), 2)
	drop := "x-nextensio-drop: " + reason + "\r\n"
	dropb := []byte(drop)
	pak = bytes.Join([][]byte{z[0], dropb}, []byte("\r\n"))
	r := regexp.MustCompile("content-length:.+")
	pak = r.ReplaceAll(pak, []byte("content-length: 0"))
	_, e := nullConn.Write(pak)
	if e != nil {
		nullConn.Close()
		nullConn = nil
	}
}
